ebos/eclinterregflows.hh

```diff
@@ -0,0 +1,362 @@
+        ///    definition array across all MPI ranks.
+        ///
+        /// \return Whether or not assignment succeeded.
+        bool assignGlobalMaxRegionID(const std::vector<std::size_t>& regID);
+
+        /// Whether or not previous read() operation succeeded.
+        bool readIsConsistent() const;
+
+        /// Serialise internal representation to MPI message buffer
+        ///
+        /// \tparam MessageBufferType Linear MPI message buffer.  API should
+        ///    be similar to Dune::MessageBufferIF
+        ///
+        /// \param[in,out] buffer Linear MPI message buffer instance.
+        ///    Function appends a partially linearised representation of
+        ///    \code *this \endcode to the buffer contents.
+        template <class MessageBufferType>
+        void write(MessageBufferType& buffer) const
+        {
+            buffer.write(this->names_.size());
+            for (const auto& name : this->names_) {
+                buffer.write(name);
+            }
+
+            for (const auto& map : this->regionMaps_) {
+                map.write(buffer);
+            }
+        }
+
+        /// Reconstitute internal object representation from MPI message
+        /// buffer
+        ///
+        /// This object (\code *this \endcode) is not usable in subsequent
+        /// calls to \code addConnection() \endcode following a call to
+        /// member function \code read() \endcode.
+        ///
+        /// \tparam MessageBufferType Linear MPI message buffer.  API should
+        ///    be similar to Dune::MessageBufferIF
+        ///
+        /// \param[in,out] buffer Linear MPI message buffer instance.
+        ///    Function reads a partially linearised representation of \code
+        ///    *this \endcode from the buffer contents and advances the
+        ///    buffer's read position.
+        template <class MessageBufferType>
+        void read(MessageBufferType& buffer)
+        {
+            const auto names = this->readNames(buffer);
+
+            if (names == this->names_) {
+                // Input stream holds the maps in expected order (common
+                // case).  Read the maps and merge with internal values.
+                for (auto& map : this->regionMaps_) {
+                    map.read(buffer);
+                }
+            }
+            else {
+                // Input stream does not hold the maps in expected order (or
+                // different number of maps).  Unexpected.  Read the values
+                // from the input stream, but do not merge with internal
+                // values.
+                auto map = EclInterRegFlowMapSingleFIP {
+                    std::vector<int>(this->numCells_, 1)
+                };
+
+                const auto numMaps = names.size();
+                for (auto n = 0*numMaps; n < numMaps; ++n) {
+                    map.read(buffer);
+                }
+
+                this->readIsConsistent_ = false;
```

Comment:
Have not seen `readIsConsistent()` being queried somewhere besides the test? Is it used? What happens if we end up in the else branch? Aren't the names determined from either the EclState or the FieldProperties and should be the same on every process?

Reply:
> Have not seen `readIsConsistent()` being queried somewhere besides the test? Is it used?

It's not, but that's (probably) a mistake.  I intended to use this flag to determine whether or not to finalise the global aggregation of the inter-region flows through a `compress()` call in `CollectToIORank`.  I may still do that.

> What happens if we end up in the else branch?

Right now, we "just" ignore the contributions from that partner MPI rank.  If we do check the state before running the global `compress()` we'll end up ignoring all contributions from all ranks.

> Aren't the names determined from either the `EclState` or the `FieldProperties` and should be the same on every process?

They're supposed to be which is why ending up in this branch is labelled as "unexpected".  We might consider stopping the simulation if this situation ever occurs.

Reply:
> > Have not seen `readIsConsistent()` being queried somewhere besides the test? Is it used?
> 
> It's not, but that's (probably) a mistake. I intended to use this flag to determine whether or not to finalise the global aggregation of the inter-region flows through a `compress()` call in `CollectToIORank`. I may still do that.
> 
> > What happens if we end up in the else branch?
> 
> Right now, we "just" ignore the contributions from that partner MPI rank. If we do check the state before running the global `compress()` we'll end up ignoring all contributions from all ranks.
> 

This will give us a rather hard time noticing this. Why not  throw and remove the rest?

Reply:
> Why not throw and remove the rest?

I suppose that would work, but we need to be mindful of the context here.

This particular code will run during `CollectToIORank::collect()`, and in that case only on the I/O rank.  Suppose that we've protected the [relevant region](https://github.com/OPM/opm-simulators/blob/b9a1270bbb90e0f40004c5fab6cbc7a1402e3e50/ebos/eclwriter.hh#L194-L205) with `OPM_BEGIN_PARALLEL_TRY_CATCH()` and `OPM_END_PARALLEL_TRY_CATCH()` and that the third of N (> 3) message happens to fail in this very unlikely way.  What will happen to the remaining messages that have not yet been received by the I/O rank and, more importantly, to their senders?  Will this deadlock the entire set of MPI processes?

Reply:
>> Will this [unlikely scenario] deadlock the entire set of MPI processes?
>
>Yep, it might deadlock as we might need to receive messages from other processes afterwards. We need to find a way throw after the [exchange](https://github.com/OPM/opm-simulators/blob/b9a1270bbb90e0f40004c5fab6cbc7a1402e3e50/ebos/collecttoiorank.cc#L1010) has completed, e.g. by querying a data member.

That's what I feared.  In that case I'm keeping the `readIsConsistent()` predicate for this purpose.  I've pushed an update in #3811 that checks `readIsConsistent()` on the I/O rank after the `CollectToIORank::collect()` call completes and communicates the failure condition to all ranks.